### ISSUE TITLE
go_path not working correctly for vendored sources

### ISSUE THREAD

**Issue by guoshimin** [bazelbuild/rules_go#1450]
We have a mono repo where the go sources have the following directory structure:
```
- go
  \- src
     \- mypkg
     |- vendor
```
`go_path` produces the following directory structure:
```
- src
  \- mypkg
  |- go
     \- src
        \- vendor
```
which prevents standard tools from finding the vendored packages.

I think this is a result of go_path using the importmap attribute of the vendored packages. We used gazelle to auto-generate the go targets, which have the importmap attributes set automatically. Not sure what's the right thing to do here.

**Comment by jayconrod** [comment]
What do the `importmap` attributes look like for libraries in your vendor directory? The logic for this is in [path.bzl:154](https://github.com/bazelbuild/rules_go/blob/dcaae02413ca60dc857e1d710061a0a8a77adf96/go/private/tools/path.bzl#L154).

* If `vendor` does not appear in `importmap`, `go_path` just places the library according to the `importpath`. This is the common case for non-vendored libraries.
* If `importmap` starts with `<workspace>/vendor` (for some external repository), `go_path` strips off the first component and places the library in `vendor`. This is the common case for external repos and repos with explicit workspace names.
* Otherwise` go_path` will place the library according to `importmap`.

I'm guessing your vendored libraries `importmap` attributes look like `go/src/vendor/...`, which will fall into the last case.

One option here is to have `go_path` strip off everything before `vendor`. There's a risk of overlapping packages at different versions, but we're already kind of doing that with the second case above.

Another option: you could use gazelle to rewrite `importmap` attributes within your vendor directory. To do so, add this directives in your go/src/vendor/BUILD.bazel file:

```
# gazelle:importmap_prefix vendor
```

Gazelle generates `importmap` attributes by concatenating this prefix with the relative path from where the prefix was set. So a library in `go/src/vendor/foo/bar` would get `importmap = "vendor/foo/bar"`.

**Comment by guoshimin** [comment]
Tried the importmap_prefix for the vendor directory with gazelle 0.11.0 and it didn't have any effect. It seems importmap_prefix gets overwritten if the vendor appears in the path here:
https://github.com/bazelbuild/bazel-gazelle/blob/97ee9d7a9b5c44a279f2df349232378a9aa803db/internal/packages/walk.go#L138

**Comment by jayconrod** [comment]
I just tried this out. I think Gazelle is generating the correct value for `importmap` but is not replacing the old value. Try removing it, then re-running Gazelle.

You can bulk-remove attributes with [buildozer](https://github.com/bazelbuild/buildtools/tree/master/buildozer).

```
$ buildozer 'remove importmap' '//go/src/vendor/...:*'
```

That attribute should probably be mergeable. I'll fix that in Gazelle.

**Comment by guoshimin** [comment]
Oh I made a stupid mistake. I tried to remove all the BUILD.bazel files in go/src/vendor and regenerate, but in that process removed go/src/vendor/BUILD.bazel too /facepalm

So this works for me. Closing the issue.

**Comment by guoshimin** [comment]
Actually there is another problem. It seems the importmap_prefix directive doesn't affect the importmaps for nested vendors. The importmaps for those still start with go/src/vendor/.

**Comment by zanes2016** [comment]
After upgrading to the **0.11.0** release, I believe I'm seeing a similar issue (hit it when we run our unittests). The `# gazelle:importmap_prefix` suggestion does not seem to work.

For some reason the extra `vendor/` in the path is added and causing problems. This was working in version **0.10.3** but no longer works in version **0.11.0**

## ERROR
```
...
cannot use &coordSrc (type *"github.com/hashicorp/serf/coordinate".Coordinate) as type *"vendor/github.com/hashicorp/serf/coordinate".Coordinate in field value
...
```

## WORKSPACE
```
go_repository(
    name = "com_github_hashicorp_serf",
    importpath = "github.com/hashicorp/serf",
    tag = "v0.8.1"
)
```

## BUILD
```
go_test(
    name = "go_default_test",
    srcs = ["nodes_test.go"],
    embed = [":go_default_library"],
    deps = [
        "//monitor/src/common/config:go_default_library",
        "@com_github_hashicorp_consul//api:go_default_library",
        "@com_github_hashicorp_serf//coordinate:go_default_library", # gazelle:importmap_prefix github.com/hashicorp
    ],
)
```

Any suggestions?

**Comment by jayconrod** [comment]
@zanes2016 I suspect more than one copy of that library is getting linked into your test in a way where the types are clashing.

Try searching for libraries with that `importpath` depended on by that target using something like `bazel query 'attr(importpath, "github.com/hashicorp/serf/coordinate", deps(//:go_default_test))`. Ideally that should give you exactly one result. If it gives multiple results, they must have distinct `importmap` attributes. Their types won't be considered compatible, so they can't be used in the same situations.

Gazelle will set `importmap` attributes automatically for libraries in vendor directories. If you want to control this explicitly, put `# gazelle:importmap_prefix github.com/hashicorp` in the top level of a build file. It won't work inside a rule. See [Directives](https://github.com/bazelbuild/bazel-gazelle#directives). I'll add some examples to this so it's clearer.

If you're still seeing an error, would you mind posting a more complete reproducible example? I tried a simple build from what you posted, but I didn't see an error.

Also, what version of Gazelle are you using with `go_repository`?

**Comment by zanes2016** [comment]
@jayconrod thanks, I'll take a look into this. 

We are using **version 0.11.0**

```
http_archive(
    name = "io_bazel_rules_go",
    url = "https://github.com/bazelbuild/rules_go/releases/download/0.11.0/rules_go-0.11.0.tar.gz",
    sha256 = "f70c35a8c779bb92f7521ecb5a1c6604e9c3edd431e50b6376d7497abc8ad3c1",
)
load("@io_bazel_rules_go//go:def.bzl", "go_repository", "go_rules_dependencies", "go_register_toolchains")
go_register_toolchains(go_version='1.10.1')
go_rules_dependencies()
```

**Comment by jayconrod** [comment]
Gazelle is versioned separately from rules_go. Currently, rules_go pulls in a somewhat older version of Gazelle to use with the legacy version of `go_repository`, but that will go away before rules_go 0.13.0.

See [`Running Gazelle with Bazel`](https://github.com/bazelbuild/bazel-gazelle#running-gazelle-with-bazel) for info on pulling in Gazelle separately.

**Comment by zanes2016** [comment]
Updated the WORKSPACE to include the latest bazel gazelle but the issue persists.

**WORKSPACE**
```
http_archive(
    name = "io_bazel_rules_go",
    url = "https://github.com/bazelbuild/rules_go/releases/download/0.11.0/rules_go-0.11.0.tar.gz",
    sha256 = "f70c35a8c779bb92f7521ecb5a1c6604e9c3edd431e50b6376d7497abc8ad3c1",
)
http_archive(
    name = "bazel_gazelle",
    url = "https://github.com/bazelbuild/bazel-gazelle/releases/download/0.11.0/bazel-gazelle-0.11.0.tar.gz",
    sha256 = "92a3c59734dad2ef85dc731dbcb2bc23c4568cded79d4b87ebccd787eb89e8d0",
)
load("@io_bazel_rules_go//go:def.bzl", "go_repository", "go_rules_dependencies", "go_register_toolchains")
go_register_toolchains(go_version='1.10.1') 
go_rules_dependencies()
load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")
gazelle_dependencies()
```

**Comment by jayconrod** [comment]
@zanes2016 Did you find out how multiple copies of that library are getting included?

**Comment by zanes2016** [comment]
@jayconrod I haven't figured out how (or if) multiple copies of that library are getting included.

In my WORKSPACE I have:
```
...
go_repository(
    name = "com_github_hashicorp_serf",
    importpath = "github.com/hashicorp/serf",
    tag = "v0.8.1"
)
...
```

If i comment out the above go_repository, the build fails with 
`no such package '@com_github_hashicorp_serf//coordinate': The repository could not be resolved`.

I believe `com_github_hashicorp_serf` conforms to the correct gazelle naming convention

**Comment by zanes2016** [comment]
@jayconrod Here's a minimum setup to reproduce the issue:

```
# ls
BUILD  WORKSPACE  main.go
```

**WORKSPACE**
```
# cat WORKSPACE
http_archive(
    name = "io_bazel_rules_go",
    url = "https://github.com/bazelbuild/rules_go/releases/download/0.11.0/rules_go-0.11.0.tar.gz",
    sha256 = "f70c35a8c779bb92f7521ecb5a1c6604e9c3edd431e50b6376d7497abc8ad3c1",
)
http_archive(
    name = "bazel_gazelle",
    url = "https://github.com/bazelbuild/bazel-gazelle/releases/download/0.11.0/bazel-gazelle-0.11.0.tar.gz",
    sha256 = "92a3c59734dad2ef85dc731dbcb2bc23c4568cded79d4b87ebccd787eb89e8d0",
)
load("@io_bazel_rules_go//go:def.bzl", "go_repository", "go_rules_dependencies", "go_register_toolchains")
go_register_toolchains(go_version='1.10.1')  # TODO: be sure to always update!!!
go_rules_dependencies()
load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")
gazelle_dependencies()

go_repository(
    name = "com_github_hashicorp_consul",
    importpath = "github.com/hashicorp/consul",
    tag = "v1.0.7"
)

go_repository(
    name = "com_github_hashicorp_serf",
    importpath = "github.com/hashicorp/serf",
    tag = "v0.8.1"
)
```

**BUILD**
```
# cat BUILD
load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library", "go_test")

go_library(
    name = "main",
    srcs = ["main.go"],
    importpath = "main",
    visibility = ["//visibility:public"],
    deps = [
      "@com_github_hashicorp_consul//api:go_default_library",
      "@com_github_hashicorp_serf//coordinate:go_default_library",
    ],
)
```

**main.go**
```
# cat main.go 
package main

import (
	"github.com/hashicorp/consul/api"
	"github.com/hashicorp/serf/coordinate"
)

var test = &api.CoordinateEntry{
	Node: "node3",
	Coord: &coordinate.Coordinate{},
}

func main() {}
```

**Environment**
```
# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.4 LTS
Release:	16.04
Codename:	xenial
#
#
# bazel version
Build label: 0.11.1
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue May 14 07:48:23 +50148 (1520362424903)
Build timestamp: 1520362424903
Build timestamp as int: 1520362424903
```

**Example:**
```
# bazel build main
Extracting Bazel installation...
..........................................
INFO: Analysed target //:main (19 packages loaded).
INFO: Found 1 target...
ERROR: /demo/BUILD:3:1: GoCompile linux_amd64_stripped/main~/main.a failed (Exit 1)
/root/.cache/bazel/_bazel_root/53a21a31163cea6582840a4862f776cf/bazel-sandbox/5205980998850882637/execroot/__main__/main.go:10:7: cannot use "github.com/hashicorp/serf/coordinate".Coordinate literal (type *"github.com/hashicorp/serf/coordinate".Coordinate) as type *"vendor/github.com/hashicorp/serf/coordinate".Coordinate in field value
GoCompile: error running compiler: exit status 1
Target //:main failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 128.690s, Critical Path: 57.43s
FAILED: Build did NOT complete successfully
```

**Comment by zanes2016** [comment]
Thanks @jayconrod for the explaination. I take it then there's no way to accomplish the build in a single step with `bazel build`?

**Comment by jayconrod** [comment]
@zanes2016 A single Bazel build would be possible, but you'd need to edit the build files to avoid the conflict. Gazelle won't be able to do it for you. In this case, you could depend on `@com_github_hashicorp_consul//vendor/github.com/hashicorp/serf/coordinate:go_default_library` directly.

In general, rules_go is the mechanism, and Gazelle is the policy. Gazelle will generate rules that do the same thing as `go build`. In this case, `go build` reports an error, and the rules generated by Gazelle give you the same error. But if you're willing to hand-write build files, you more flexibility, but it's a lot more work.

**Comment by zanes2016** [comment]
@jayconrod Thanks! 

As a side note, any idea why we run into this issue with version *0.11.0* but not version *0.10.3*?

**Comment by jayconrod** [comment]
@zanes2016 If I remember correctly, we added `importmap` between 0.10 and 0.11. Before that, we provided the compiler and linker with the dependencies that were asked for, and if multiple packages had the same `importpath`, the one that got linked was the one that the compiler or linker found first. We had a number of bizarre link-time and run-time errors reported because of this. You might end up linking a different version of a library that you compiled against.

In 0.11, we added `importmap`. Gazelle generates `importmap` attributes automatically for vendored libraries, and we started warning when libraries have the same `importmap` (it will be an error in the future).
